# SNLP Assignment 2

Name 1: Zoha Zehra<br/>
Student id 1: 7061946<br/>
Email 1: zoze00001@stud.uni-saarland.de<br/>


Name 2: Daniya Niazi<br/>
Student id 2: 7062343<br/>
Email 2: dani00003@stud.uni-saarland.de<br/> 

Name 3: Aleksey Morshnev<br/>
Student id 3:7042691<br/>
Email 3: almo00008@stud.uni-saarland.de<br/> 

**Instructions:** Read each question carefully. <br/>
Make sure you appropriately comment your code wherever required. Your final submission should contain the completed Notebook. There is no need to submit the data files. <br/>
Upload the zipped folder on CMS. Please follow the naming convention of **Name1_studentID1_Name2_studentID2_Name3_studentID3.zip**. Make sure to click on "Turn-in" (or the equivalent on CMS) after you upload your submission, otherwise the assignment will not be considered as submitted. Only one member of the group should make the submisssion.

---

## Exercise 1 - Probability theory (0.5 + 0.5 = 1 point)

Consider a fair 6-sided die whose sides are numbered from 1 to 6 and each die roll is independent of the other rolls. In an experiment that consists of rolling the die twice, the following events can be defined:

    A: The sum of the two outcomes is at least 10
    B: At least one of the two rolls resulted in 6

a. Compute the probabilities $P(A)$, and $P(B)$.

b. Is event A independent of event B?

You can leave the solution as a fraction.

---

#### Solution 1

#### Part (a)

(1,1) 
(1,2)
(1,3)
(1,4)
(1,5)
(1,6)

(2,1)
(2,2)
(2,3)
(2,4)
(2,5)
(2,6)

(3,1)
(3,2)
(3,3)
(3,4)
(3,5)
(3,6)

(4,1)
(4,2)
(4,3)
(4,4)
(4,5)
(4,6)

(5,1)
(5,2)
(5,3)
(5,4)
(5,5)
(5,6)

(6,1)
(6,2)
(6,3)
(6,4)
(6,5)
(6,6)

Total Outcome: 36 


A: The sum of the two outcomes is at least 10

P(A) = 6/36 = 1/6

B: At least one of the two rolls resulted in 6

P(B) = 11/36
 

#### Part (b)

The favorable outcomes for event A and B are:
(4, 6), (5, 6), (6, 4), (6, 5), (6, 6)

P(A and B) = 5/11

P(A) × P(B) = (1/6) x (11/36) = 11/216

Since P(A and B) is not equal to P(A) × P(B), events A and B are not independent.

## Exercise 2 - Bayes theorem (0.5 + 1 = 1.5 points)

Bayes' theorem is a fundamental concept in probability theory used to update the probability of a hypothesis based on new evidence. It states that the probability of a hypothesis (H) given some evidence (E) is proportional to the probability of the evidence given the hypothesis, multiplied by the prior probability of the hypothesis, divided by the probability of the evidence regardless of the hypothesis. Mathematically,

$$ P(H|E) = \frac{P(E|H)P(H)}{P(E)} $$

Where,
- $P(H∣E)$ is the posterior probability of hypothesis H given evidence E.
- $P(E∣H)$ is the probability of observing evidence E given that the hypothesis H is true.
- $P(H)$ is the prior probability of hypothesis H being true.
- $P(E)$ is the probability of observing evidence E.


Now, Imagine that all of the blogs online were generated by some Large Language Model (LLM). Of these blogs, $80\%$ were generated using GPT-4, $15\%$ of them were generated using LLAMA 3 and $5\%$ were generated using Mistral. About $4\%$ of the text generated by GPT-4 seem to have counterfactual statements. For LLAMA 3, this is about $6\%$ and for Mistral, this is about $9\%$.

a. If a blog post is randomly selected from these online blogs, what is the probability that it was generated by LLAMA 3?

b. If a randomly selected blog post contained a counterfactual statement, find the probability that it was generated using GPT-4.

---

#### Solution 2

#### Part (a)

P(Generated by LLAMA 3) = 0.15

#### Part (b)

H: The blog post was generated using GPT-4

E: The blog post contains a counterfactual statement

P(H) = 0.8

P(E|H) = 0.04

P(E) = P(E|H) x P(H) + P(E|!H) x P(!H)

P(E|H) = 0.04

P(H) = 0.8

P(!H) = 1 - 0.8

P(!H) = 0.2

P(E|!H) = (0.06 * 0.15 + 0.09 * 0.05) / 0.2

P(E|!H) = 0.0675

P(E) = (0.8 x 0.04) / 0.0675 x 0.2 = 0.0495

P(H|E)  = (0.04 * 0.8) / 0.0495 

P(H|E) =  0.6464


## Exercise 3 - Miller's model (0.5 + 1 + 0.5 = 2 points)

Miller's model introduced in the lecture simulates the process of random text generation at the character level. It is a memoryless model which means that the generation at each time step is independent of what has been generated so far. Imagine that there is a monkey typing on a computer keyboard. The keyboard has only 27 keys: a to z and a spacebar. Assume that the probability of each character is distributed according to its occurrence in the English language. We call a sequence of characters separated by white space as a word.

a. What is the probability that the monkey will type the word _'ear'_ on the keyboard? Hint: See the `exercise_3.py` file.

b. Complete the `compute_perplexity()` function in the `exercise_3.py` file to compute the perplexity of the Miller's model for the text: "hello i am a monkey". 

c. What would happen to the perplexity if we were to add a few more whitespace in between the words?

---

#### Solution 3

As per millers model the random generation is at character level. </br>

P('ear') = P('e') * P('a') * P('r') </br>

P('ear') = 0.08929 * 0.06798 * 0.06065 = 0.00036814150922999996

In [1]:
# Solution 3.b

# DO NOT MODIFY THIS PART
# Your code should be placed in exercise_3.py
import exercise_3
from importlib import reload

reload(exercise_3)

model = exercise_3.MillersModel()

text = "hello i am a monkey"

print("Perplexity: ", model.compute_perplexity(text))

Perplexity:  17.628396499801106


In [2]:
# Solution 3.c

## lets add more space to text


text = "he llo i am a mon key"

print("Perplexity with spaces: ", model.compute_perplexity(text))

text = "he ll o i am a m on ke y"

print("Perplexity with more spaces: ", model.compute_perplexity(text))

Perplexity with spaces:  15.634876233263183
Perplexity with more spaces:  13.558257950994317


As I'm increasing the spaces the perplexity is getting decreasing indiciating higher perfomace at non meaningful arranged sequence of words. We have passed a meaningful sentence but 17.62 value is indicating that this model is still uncertain even in correct sequence of words

## Exercise 4 - Sampling in Language models (1 + 2 + 0.5 + 0.5 =4 points)

Neural Networks are quite effective as generative models. Huggingface Transformers library can be used to easily get started with training and using these models. Take a look at the [Huggingface transformers quick tour](https://huggingface.co/docs/transformers/quicktour) to get acquainted with the transformers library. (Recommended: [Chapter 1](https://huggingface.co/learn/nlp-course/chapter1/1) and [Chapter 2](https://huggingface.co/learn/nlp-course/chapter2/1) of the Huggingface NLP Course)

a. Read up on GPT2 [[1]](https://huggingface.co/openai-community/gpt2) [[2]](https://huggingface.co/docs/transformers/model_doc/gpt2). Implement the `forward()` function in the `GPT2Model` class in `exercise_4.py` file to take input text sequence and return the probability distribution of the predicted next token. The general steps would be: tokenize and encode the input text, feed the input to the model to get the logits (Note: You might have to take only the last logit which is for the predicted next token in the sequence.), and apply softmax function to convert it to a probability distribution.

b. We can use different types of sampling strategies while selecting the next token for generating a text. Some common approaches are: Greedy sampling, random sampling, beam search, etc.

   - Greedy sampling
      In greedy sampling, the model selects the word with the highest probability based on the given context. It prioritizes the most likely next word based on the model's predictions, without considering other possibilities. This method often produces coherent and fluent text but can lead to repetitive or predictable outputs, as it doesn't explore alternative word choices.

   - Random sampling
      In random sampling, the next token is randomly sampled from the output probability distribution. This can make the generation non-deterministic and unpredictable but doesn't usually fall into repetitions.

   Implement the `greedy_sample()` and `random_sample()` functions in `GPT2Model` class in `exercise_4.py` file for Greedy sampling and random sampling respectively. For each of them, generate a continuation of the sentence given below.

c. Describe the output with each sampling method. Which one would you generally prefer?

d. Another approach for generating from language models is using Beam search. Describe how the beam search works (2 sentences). Does it solve any problems of the previous sampling methods?

---

In [3]:
# install necessary packages
# !conda install transformers evaluate datasets

In [4]:
# Solution 4

# DO NOT MODIFY THIS PART
# Your code should be placed in exercise_4.py
import exercise_4
from importlib import reload

reload(exercise_4)

model = exercise_4.GPT2Model()

In [5]:
predictions = model.forward("Alice was beginning to get very tired of sitting by her sister on the bank ")
predictions.size()

torch.Size([1, 50257])

In [6]:
# Greedy sampling
n_tokens = 10

text = "Alice was beginning to get very tired of sitting by her sister on the bank "

for _ in range(n_tokens):
    text += model.greedy_sample(text)

print("Greedy sampled output: ", text)

Greedy sampled output:  Alice was beginning to get very tired of sitting by her sister on the bank   and she was starting to get tired of sitting


In [7]:
# Random sampling
n_tokens = 10

text = "Alice was beginning to get very tired of sitting by her sister on the bank "

for _ in range(n_tokens):
    text += model.random_sample(text)

print("Random sampled output: ", text)

Random sampled output:  Alice was beginning to get very tired of sitting by her sister on the bank urchin and while she tried to talk to her


#### 4.c

Greedy sample use the information that Alice is sitting and now she is tired of sitting ( knowing that she is sitting ) however random sample text do not make any meaning with the alice being sitting near the bank and trying to get something done.

I would prefer greedy sample for text generation scenarios and for random text generation randam sample would work fine.

#### 4.d Beam search

Beam search works by maintaing history of n best candidates ( n is the beam width ) for the resulting output text. This could provide the results from n best combination instead if focusing on only 1 best predictions and never explored the outcome of other candidtate further.

Previous sampling method could lead to single dead end where further text sense is lost. In case of beam search we can back track to second best search


## Exercise 5 - Computing Perplexity (0.5 + 1 = 1.5 points)

Huggingface provides the `evaluate` library that includes various metrics used in machine learning along with perplexity that we use in Language modeling. We can use that library to quickly compute the perplexity of a language model. There is also the `datasets` library from Huggingface that allows to easily access datasets for many different tasks.

a. Read the [metric card for perplexity](https://huggingface.co/spaces/evaluate-metric/perplexity) and compute the perplexity for the following text using GPT2.

b. Load the `alice_in_wonderland.txt` file using Hugginface `datasets` library and compute the perplexity for it using GPT2. Check the [documentations](https://huggingface.co/docs/datasets/index) for datasets library.

---

In [8]:
# Solution 5.a
from evaluate import load

texts = ["Eureka! I have found it!"]

# your code here
perplexity = load("perplexity", module_type="metric")
results = perplexity.compute(predictions=texts, model_id='gpt2')


print("Perplexity: ",results['mean_perplexity'] )

  0%|          | 0/1 [00:00<?, ?it/s]

Perplexity:  24.384519577026367


In [9]:
# Solution 5.b
from transformers import GPT2Tokenizer
from datasets import load_dataset
from evaluate import load

# Load the tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('gpt2')

# Load the dataset
dataset = load_dataset('text', data_files={'train': 'alice_in_wonderland.txt'}, split='train')

# Function to check token length
def check_token_length(text):
    tokens = tokenizer.encode(text, add_special_tokens=True)
    if len(tokens) < 1:  # Checks if the token length is less than 1
        return False
    return True

# Filter and extract text data ensuring each line results in at least one token
valid_texts = []
for item in dataset:
    text = item['text'].strip()
    if text and check_token_length(text):
        valid_texts.append(text)

perplexity = load("perplexity", module_type="metric")
results = perplexity.compute(predictions=valid_texts, model_id='gpt2')

# map the perplexities to the corresponding text
# perplexities = dict(zip(valid_texts, results['perplexities']))
# print(perplexities)

print("Mean Perplexity: ", results['mean_perplexity'])

  0%|          | 0/2 [00:00<?, ?it/s]

Mean Perplexity:  633.9214832999489


## Bonus

## Rejection Sampling (2 points)

Rejection sampling is another approach we can use to generate the next token from the language model. In this approach, we randomly sample a token from the list and accept it if it's probability is greater than some random number and reject it otherwise.

1. Generate a random number, r.
2. Randomly sample a token from the output probability (say token at index i with probability p).
3. Is r < p?

    a. If yes, then accept i as the next token.
    
    b. If no, then go back to step 1.

Complete the `rejection_sample()` function in the `GPT2Model` class in `exercise_4.py`. How does the generated output differ from the greedy and random sampling?

---

Solution Bonus:

Rejection Sampling method introduces a balance between randomness and probability weighted choices. Output is a compromise between random sampling and greedy sampling.

In [10]:
# Random sampling
n_tokens = 10

text = "Alice was beginning to get very tired of sitting by her sister on the bank "

for _ in range(n_tokens):
    text += model.rejection_sample(text)

print("Rejection sampled output: ", text)

Rejection sampled output:  Alice was beginning to get very tired of sitting by her sister on the bank   and was starting to think about her own future


## Perplexity with different languages (0.5 + 0.5 = 1 point)

a. In exercise 5.a., what would happen if we were to use a different language other than English? How would the perplexity be affected? Test this with another language of your choice.

b. Are there any pre-trained model for the language you selected? Try with that model to see if you can get a better perplexity score. If there are no pre-trained LM for that language, try with a multilingual LM and see if the perplexity score improves.

Solution:

a. GPT-2 was primarily trained on English text, so predicting the next token in a different language results in a much higher perplexity score.

In [11]:
from evaluate import load

# Load the perplexity metric
perplexity = load("perplexity", module_type="metric")

# Uzbek text to compute perplexity
uzbek_text = ["Evrika! Men topdim!"]

# Compute the perplexity using the GPT-2 model
results = perplexity.compute(predictions=uzbek_text, model_id='gpt2')

# Print the perplexity result
print("Perplexity for Uzbek text: ", results['mean_perplexity'])


  0%|          | 0/1 [00:00<?, ?it/s]

Perplexity for Uzbek text:  41356.34375


Solution:

b. There is a pretrained model for Uzbek language called "mGPT-1.3B-uzbek". Using it, the perplexity score decresed from 41356 to 301.

In [12]:
from evaluate import load

# Load the perplexity metric
perplexity = load("perplexity", module_type="metric")

# Uzbek text to compute perplexity
uzbek_text = ["Evrika! Men topdim!"]

# Compute the perplexity using the GPT-2 model
results = perplexity.compute(predictions=uzbek_text, model_id='ai-forever/mGPT-1.3B-uzbek')

# Print the perplexity result
print("Perplexity for Uzbek text: ", results['mean_perplexity'])

/home/reni/miniconda3/envs/hws_ws23/lib/python3.10/site-packages/torch/_utils.py:831: UserWarning: TypedStorage is deprecated. It will be removed in the future and UntypedStorage will be the only storage class. This should only matter to you if you are using storages directly.  To access UntypedStorage directly, use tensor.untyped_storage() instead of tensor.storage()
  return self.fget.__get__(instance, owner)()


  0%|          | 0/1 [00:00<?, ?it/s]

Perplexity for Uzbek text:  301.0086364746094
